src/Dictionaries/HashedDictionary.cpp

```diff
@@ -35,16 +41,151 @@ namespace ErrorCodes
     extern const int BAD_ARGUMENTS;
     extern const int DICTIONARY_IS_EMPTY;
     extern const int UNSUPPORTED_METHOD;
+    extern const int LOGICAL_ERROR;
 }
 
-template <DictionaryKeyType dictionary_key_type, bool sparse>
-HashedDictionary<dictionary_key_type, sparse>::HashedDictionary(
+template <DictionaryKeyType dictionary_key_type, bool sparse, bool sharded> class HashedDictionary;
+
+/// Implementation parallel dictionary load for SHARDS
+template <DictionaryKeyType dictionary_key_type, bool sparse, bool sharded>
+class ParallelDictionaryLoader : public boost::noncopyable
+{
+    using HashedDictionary = HashedDictionary<dictionary_key_type, sparse, sharded>;
+    using Queue = ConcurrentBoundedQueue<Block>;
+
+public:
+    explicit ParallelDictionaryLoader(HashedDictionary & dictionary_)
+        : dictionary(dictionary_)
+        , shards(dictionary.configuration.shards)
+        , simple_key(dictionary.dict_struct.getKeysSize() == 1)
+        , pool(shards)
+        , shards_queues(shards)
+    {
+        UInt64 backlog = dictionary.configuration.shard_load_queue_backlog;
+        LOG_TRACE(dictionary.log, "Will load the dictionary using {} threads (with {} backlog)", shards, backlog);
+
+        shards_slots.resize(shards);
+        std::generate(shards_slots.begin(), shards_slots.end(), [n = 0]() mutable { return n++; });
+
+        for (size_t shard = 0; shard < shards; ++shard)
+        {
+            shards_queues[shard].emplace(backlog);
+            pool.scheduleOrThrowOnError([this, shard, thread_group = CurrentThread::getGroup()]
+            {
+                if (thread_group)
+                    CurrentThread::attachToIfDetached(thread_group);
+                setThreadName("HashedDictLoad");
+
+                threadWorker(shard);
+            });
+        }
+    }
+
+    void addBlock(Block block)
+    {
+        Blocks shards_blocks(shards);
+        for (size_t shard = 0; shard < shards; ++shard)
+            shards_blocks[shard] = block.cloneEmpty();
+
+        IColumn::Selector selector = createShardSelector(block, shards_slots);
+        splitBlock(selector, block, shards_blocks);
+
+        for (size_t shard = 0; shard < shards; ++shard)
+        {
+            if (!shards_queues[shard]->push(std::move(shards_blocks[shard])))
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Could not push to shards queue #{}", shard);
```

Comment:
`ConcurrentBoundedQueue::push` never returns false, I guess it returns `bool` to make the interface similar to `tryPush`, although not sure that this is good, but anyway, to avoid touching that file here - throw `LOGICAL_ERROR`

And actually that fact, that `ConcurrentBoundedQueue::push` always return `true` does not looks correct, yes it does not accept timeout, but in case of the queue is cleared (`clearAndFinish`) it should return false.

Reply:
- #44881